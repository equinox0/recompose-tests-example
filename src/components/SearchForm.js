import React from "react";
import { withHandlers } from "recompose";
import PropTypes from "prop-types";

const SearchForm = ({ onChange, onSubmit, values }) => (
    <form onSubmit={onSubmit}>
        <input name="vehicleId" onChange={onChange} type="number" value={values["vehicleId"]} />
        <button type="submit">Szukaj</button>
    </form>
);

SearchForm.propTypes = {
    onChange: PropTypes.func.isRequired,
    onSubmit: PropTypes.func.isRequired,
    values: PropTypes.shape({
        vehicleId: PropTypes.number.isRequired
    }).isRequired
};

const enhance = withHandlers({
    onChange: ({ onChange, values }) => ({ target }) => {
        onChange({ ...values, [target.name]: Number(target.value) });
    },
    onSubmit: ({ onSubmit }) => event => {
        event.preventDefault();
        onSubmit();
    }
});

export { SearchForm, enhance };
export default enhance(SearchForm);
