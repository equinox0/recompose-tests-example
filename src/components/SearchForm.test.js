import { mount } from "enzyme";
import React from "react";
import { SearchForm, enhance } from "./SearchForm";

describe("<SearchForm/>", () => {
    describe("component", () => {
        const setup = overriddenProps => {
            const props = {
                onChange: jest.fn(),
                onSubmit: jest.fn(),
                values: { vehicleId: 0 },
                ...overriddenProps
            };
            const wrapper = mount(<SearchForm {...props} />);
            return { wrapper };
        };

        it("should display search form", () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });

        it("should call onChange", () => {
            const onChange = jest.fn();
            const { wrapper } = setup({ onChange });
            wrapper.find('input[name="vehicleId"]').simulate("change");
            expect(onChange).toBeCalled();
        });

        it("should call onSubmit with button", () => {
            const onSubmit = jest.fn();
            const { wrapper } = setup({ onSubmit });
            wrapper.find('button[type="submit"]').simulate("submit");
            expect(onSubmit).toBeCalled();
        });

        it("should call onSubmit with form submit", () => {
            const onSubmit = jest.fn();
            const { wrapper } = setup({ onSubmit });
            wrapper.find("form").simulate("submit");
            expect(onSubmit).toBeCalled();
        });
    });

    describe("enhance", () => {
        const setup = overriddenProps => {
            const props = {
                onChange: jest.fn(),
                onSubmit: jest.fn(),
                values: { vehicleId: 0 },
                ...overriddenProps
            };
            const MockComponent = () => null;
            const EnhancedComponent = enhance(MockComponent);
            const wrapper = mount(<EnhancedComponent {...props} />);
            return { MockComponent, wrapper };
        };

        it("should call onChange with new form values", () => {
            const onChange = jest.fn();
            const { MockComponent, wrapper } = setup({ onChange });
            const props = wrapper.find(MockComponent).props();
            props.onChange({ target: { name: "vehicleId", value: "666" } });
            expect(onChange).toBeCalledWith({ vehicleId: 666 });
        });

        it("should prevent default submit and call onSubmit", () => {
            const onSubmit = jest.fn();
            const event = {
                preventDefault: jest.fn()
            };
            const { MockComponent, wrapper } = setup({ onSubmit });
            const props = wrapper.find(MockComponent).props();
            props.onSubmit(event);
            expect(event.preventDefault).toBeCalled();
            expect(onSubmit).toBeCalled();
        });
    });
});
