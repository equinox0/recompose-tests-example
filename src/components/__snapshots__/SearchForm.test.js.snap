// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm/> component should display search form 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="vehicleId"
    onChange={[Function]}
    type="number"
    value={0}
  />
  <button
    type="submit"
  >
    Szukaj
  </button>
</form>
`;
